mkosi/__init__.py

```diff
@@ -645,19 +647,152 @@ def gen_kernel_images(state: MkosiState) -> Iterator[tuple[str, Path]]:
         yield kver.name, kimg
 
 
+def module_path_to_name(path: Path) -> str:
+    return path.name.partition(".")[0]
+
+
+def resolve_module_dependencies(root: Path, kver: str, modules: Sequence[str]) -> tuple[list[Path], list[Path]]:
+    """
+    Returns a tuple of lists containing the paths to the module and firmware dependencies of the given list
+    of module names (including the given module paths themselves).
+    """
+    allmodules = root.joinpath("usr/lib/modules").joinpath(kver).joinpath("kernel").glob("**/*.ko*")
+    nametofile = {module_path_to_name(m): m for m in allmodules}
+
+    # We could run modinfo once for each module but that's slow. Luckily we can pass multiple modules to
+    # modinfo and it'll process them all in a single go. We get the modinfo for all modules to build two maps
+    # that map the path of the module to its module dependencies and its firmware dependencies respectively.
+    info = run(["modinfo", "--basedir", root, "--set-version", kver, "--null", *nametofile.keys()],
+               text=True, stdout=subprocess.PIPE).stdout
+
+    moddep = {}
+    firmwaredep = {}
+
+    name = ""
+    depends = []
+    firmware = []
+    for line in info.split("\0"):
+        key, sep, value = line.partition(":")
+        if not sep:
+            key, sep, value = line.partition("=")
+
+        if key in ("depends", "softdep"):
+            for d in value.strip().split(","):
+                if not d:
+                    continue
+
+                if d not in nametofile:
+                    logging.warning(f"{d} is a dependency of {name} but is not installed, ignoring ")
+                    continue
+
+                depends.append(d)
+
+        if key == "firmware":
+            for f in root.joinpath("usr/lib/firmware").glob(f"{value.strip()}*"):
+                firmware.append(f)
+
+        if key == "filename":
+            if name:
+                moddep[name] = depends
+                firmwaredep[name] = firmware
+
+                depends = []
+                firmware = []
+
+            name = module_path_to_name(Path(value))
+
+    # Make sure we add the last module as well.
+    moddep[name] = depends
+    firmwaredep[name] = firmware
+
+    todo = [*modules]
+    mods = set()
+    firmware = set()
+
+    while len(todo) > 0:
+        m = todo.pop()
+        if not m or m in mods:
+            continue
+
+        mods.add(m)
+        todo += moddep.get(m, [])
+        firmware.update(firmwaredep.get(m, []))
+
+    return [nametofile[m] for m in mods], list(firmware)
```

Comment:
Not sure if this is relevant for reproducibility, but unlike dicts, sets are not sorted in insertion order so the `list(firmware)` will have arbitrary order. If that's relevant this should be `sorted(firmware)` instead or a dict used instead of a set by just using the dict keys as a sorted keys with some dummy `None` values. Dito for mods.